rust/connlib/clients/shared/src/lib.rs

```diff
@@ -179,13 +122,41 @@ async fn connect<CB>(
             .build(),
     );
 
-    let mut eventloop = Eventloop::new(tunnel, portal);
+    let mut eventloop = Eventloop::new(tunnel, portal, rx);
+
+    std::future::poll_fn(|cx| eventloop.poll(cx))
+        .await
+        .map_err(Error::PortalConnectionFailed)?;
+
+    Ok(())
+}
 
-    match std::future::poll_fn(|cx| eventloop.poll(cx)).await {
-        Ok(never) => match never {},
-        Err(e) => {
-            tracing::error!("Eventloop failed: {e}");
-            let _ = callbacks.on_disconnect(&Error::PortalConnectionFailed); // TMP Error until we have a narrower API for `onDisconnect`
+/// A supervisor task that handles, when [`connect`] exits.
+async fn connect_supervisor<CB>(connect_handle: JoinHandle<Result<(), Error>>, callbacks: CB)
+where
+    CB: Callbacks,
+{
+    match connect_handle.await {
+        Ok(Ok(())) => {
+            tracing::info!("connlib exited gracefully");
+        }
+        Ok(Err(e)) => {
+            tracing::error!("connlib failed: {e}");
+            let _ = callbacks.on_disconnect(&e);
         }
+        Err(e) => match e.try_into_panic() {
```

Comment:
`tokio` catches any panics in the `connect` task for us and returns a `JoinError` from the handle. We can pick that up again here without needing to install our own panic handler.